compiler/optimizations/replaceArrayAccessesWithRefTemps.cpp

```diff
@@ -0,0 +1,258 @@
+// a single array access stored into a 'ref' temporary, then reuse the temp.
+// The array access ('this') function can be expensive, so computing it only
+// once and reusing the result can save time.
+//
+// TODO: I think this only works with serial 'for' loops right now.  Extend
+//       it to apply to forall loops.
+// TODO: See if the "removable array access" pragma can go away.
+// TODO: Make sure this is OK to do with all array types and disable it for
+//       any types where it is not OK.
+// TODO: This optimization currently needs to run before the lowerIterators
+//       pass.  It would be good to fix it up to allow it to run later,
+//       as most other optimization passes do.  The challenges are that it
+//       would need to handle CForLoops instead of just ForLoops, and it
+//       needs to be able to find the index variable of the loop, which is
+//       not as obvious with a CForLoop.
+//
+// For example:
+//   var A, B: [1..n] real;
+//   for i in 1..n {
+//     A[i] = A[i] * A[i];
+//     B[i] = A[i] + f(A[i]);
+//   }
+//
+// Becomes:
+//   var A, B: [1..n] real;
+//   for i in 1..n {
+//     ref tmp = A[i];
+//     tmp = tmp * tmp;
+//     B[i] = tmp + f(tmp);
+//   }
+//
+void replaceArrayAccessesWithRefTemps() {
+  if (!fReplaceArrayAccessesWithRefTemps)
+    return;
+  forv_Vec(BlockStmt, block, gBlockStmts) {
+    if (ForLoop* forLoop = toForLoop(block)) {
+      std::vector<ContextCallExpr*> allContextCalls;
+      SymExpr* loopIdx = forLoop->indexGet();
+      CallExpr* indexMove = NULL;
+      Symbol* indexVar = NULL;
+      std::map<Symbol*, std::vector<ContextCallExpr*> > arrayAccessMap;
+      std::vector<BaseAST*> asts;
+      collect_asts(forLoop, asts);
+      for_vector(BaseAST, ast, asts) {
+        if (CallExpr* call = toCallExpr(ast)) {
+          // find the move that stores the for loop's index variable into
+          // the user-level index variable.  A ref var pointing at array
+          // elements will be inserted and initialized after this move.
+          if (call->isPrimitive(PRIM_MOVE)) {
+            if (SymExpr* rhs = toSymExpr(call->get(2))) {
+              if (rhs->var == loopIdx->var) {
+                if (toSymExpr(call->get(1))->var->hasFlag(FLAG_TEMP)) {
+                  loopIdx = toSymExpr(call->get(1));
+                } else {
+                  assert(indexMove == NULL && indexVar == NULL);
+                  indexMove = call;
+                  indexVar = toSymExpr(call->get(1))->var;
+                }
+              }
+            }
+          }
+        }
+      }
+      if (!indexMove) {
+        // If we couldn't find an expected index move, skip the optimization
+        if (DEBUG_RAAWRT) {
+          printf("%s:%d: Couldn't find index move.  "
+                 "Not replacing accesses in loop\n",
+                 forLoop->fname(), forLoop->linenum());
+        }
+        continue;
+      }
+      for_vector(BaseAST, astNode, asts) {
```

Comment:
This only wants to look at the ContextCallExprs that are within the forLoop it is looking at, not all ContextCallExprs throughout the program.